File: intelligence/kt/src/main/java/ai/flower/intelligence/Typing.kt
// Copyright 2025 Flower Labs GmbH. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

package ai.flower.intelligence

import kotlinx.serialization.Serializable
import kotlinx.serialization.SerialName

@Serializable
data class Progress(
    val totalBytes: Int? = null,
    val loadedBytes: Int? = null,
    val percentage: Double? = null,
    val description: String? = null
)

@Serializable
data class Message(
    val role: String,
    val content: String,
    val toolCalls: List<ToolCall>? = null
)

typealias ToolCall = Map<String, ToolCallDetails>

@Serializable
data class ToolCallDetails(
    val name: String,
    val arguments: Map<String, String>
)

@Serializable
data class ToolParameterProperty(
    val type: String,
    val description: String,
    val `enum`: List<String>? = null
)

@Serializable
data class ToolFunctionParameters(
    val type: String,
    val properties: Map<String, ToolParameterProperty>,
    val required: List<String>
)

@Serializable
data class ToolFunction(
    val name: String,
    val description: String,
    val parameters: ToolFunctionParameters
)

@Serializable
data class Tool(
    val type: String,
    val function: ToolFunction
)

@Serializable
data class StreamEvent(
    val chunk: String
)

data class ChatOptions(
    var model: String? = null,
    var temperature: Float? = null,
    var maxCompletionTokens: Int? = null,
    var stream: Boolean = false,
    var onStreamEvent: ((StreamEvent) -> Unit)? = null,
    var tools: List<Tool>? = null,
    var forceRemote: Boolean = false,
    var forceLocal: Boolean = false,
    var encrypt: Boolean = false
)

@Serializable
data class ChoiceMessage(
    val role: String,
    val content: String? = null,
    @SerialName("tool_calls")
    val toolCalls: List<ToolCall>? = null
)

@Serializable
data class Choice(
    val index: Int,
    val message: ChoiceMessage
)

@Serializable
data class StreamChoice(
    val index: Int,
    val delta: DeltaMessage
)

@Serializable
data class DeltaMessage(
    val content: String,
    val role: String
)

@Serializable
data class Usage(
    @SerialName("completion_tokens")
    val completionTokens: Int,
    @SerialName("prompt_tokens")
    val promptTokens: Int,
    @SerialName("total_tokens")
    val totalTokens: Int
)

@Serializable
data class ChatCompletionsRequest(
    val model: String,
    val messages: List<Message>,
    val temperature: Float? = null,
    @SerialName("max_completion_tokens")
    val maxCompletionTokens: Int? = null,
    val stream: Boolean? = null,
    val tools: List<Tool>? = null,
    val encrypt: Boolean? = null
)

@Serializable
data class ChatCompletionsResponse(
    val `object`: String,
    val created: Int,
    val model: String,
    val choices: List<Choice>,
    val usage: Usage
)

@Serializable
data class ModelListResponse(
    val `object`: String,
    val data: List<ModelData>
)

@Serializable
data class ModelData(
    val id: String,
    val `object`: String,
    val created: Int,
    @SerialName("owned_by")
    val ownedBy: String
)

File: framework/py/flwr/supercore/typing.py
# Copyright 2026 Flower Labs GmbH. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Flower type definitions for infrastructure."""


from dataclasses import dataclass

from flwr.supercore.constant import DEFAULT_SIMULATION_BACKEND_NAME


@dataclass
class SimulationClientResources:
    """Resource configuration for the ClientApp."""

    num_cpus: float | None = None
    num_gpus: float | None = None

    def __post_init__(self) -> None:
        """Validate client resources configuration."""
        if self.num_cpus is not None and not isinstance(self.num_cpus, (int, float)):
            raise ValueError("client-resources.num-cpus must be a number (int/float).")
        if self.num_gpus is not None and not isinstance(self.num_gpus, (int, float)):
            raise ValueError("client-resources.num-gpus must be a number (int/float).")


@dataclass
class SimulationInitArgs:
    """Initialization arguments for the simulation."""

    num_cpus: int | None = None
    num_gpus: int | None = None
    logging_level: str | None = None
    log_to_drive: bool | None = None

    def __post_init__(self) -> None:
        """Validate initialization arguments."""
        if self.num_cpus is not None and not isinstance(self.num_cpus, int):
            raise ValueError("init-args.num-cpus must be an integer.")
        if self.num_gpus is not None and not isinstance(self.num_gpus, int):
            raise ValueError("init-args.num-gpus must be an integer.")
        if self.logging_level is not None and not isinstance(self.logging_level, str):
            raise ValueError("init-args.logging-level must be a string.")
        if self.log_to_drive is not None and not isinstance(self.log_to_drive, bool):
            raise ValueError("init-args.log-to-drive must be a boolean.")


@dataclass
class SimulationBackendConfig:
    """Backend configuration for the simulation."""

    client_resources: SimulationClientResources | None = None
    init_args: SimulationInitArgs | None = None
    name: str = DEFAULT_SIMULATION_BACKEND_NAME

    def __post_init__(self) -> None:
        """Validate backend configuration."""
        if not isinstance(self.name, str):
            raise ValueError("backend.name must be a string.")


@dataclass
class SuperLinkSimulationOptions:
    """Options for local simulation."""

    num_supernodes: int
    backend: SimulationBackendConfig | None = None

    def __post_init__(self) -> None:
        """Validate simulation options."""
        if not isinstance(self.num_supernodes, int):
            raise ValueError(
                "Invalid simulation options: num-supernodes must be an integer."
            )


@dataclass
class SuperLinkConnection:
    """SuperLink connection configuration for CLI commands."""

    name: str
    address: str | None = None
    root_certificates: str | None = None
    insecure: bool | None = None
    enable_account_auth: bool | None = None
    options: SuperLinkSimulationOptions | None = None
